kickstarter/android-oss
# NT-2059: Update stripe sdk (payments) from 12.0.1 to 16.10.0

# 📲 What

- Update stripe sdk from 12.0.1 to 16.10.0

# 🤔 Why
<img width="997" alt="Screen Shot 2021-06-21 at 2 47 39 PM" src="https://user-images.githubusercontent.com/4083656/122832079-ada77d00-d29f-11eb-8264-83e554afa98d.png">

- See https://github.com/stripe/stripe-android/blob/master/MIGRATING.md
- See https://github.com/stripe/stripe-android/blob/master/CHANGELOG.md

# 🛠 How
- `Card` type from Stripe dependency is now internal and not accessible, substitute over `CardParams`
- `Token` constructor is now internal no longer available for testing
- [CardParams](https://github.com/stripe/stripe-android/blob/master/payments-core/src/main/java/com/stripe/android/model/CardParams.kt) internal variables for cards like `number`, `cvc`, `expMonth`, `expYear` no available as they were previously`. 
- Any validation related to those previously mentioned fields takes place on stipe end see [CardInputWidget](https://github.com/stripe/stripe-android/blob/9799aa052eddc01c457b4d6ee74a7dac6dabaed3/payments-core/src/main/java/com/stripe/android/view/CardInputWidget.kt#L219)
- Remove any use for the deprecated class [CardUtils](https://github.com/stripe/stripe-android/blob/master/payments-core/src/main/java/com/stripe/android/CardUtils.kt)

# 👀 See
- Trying to pay in a non USA project, the not available payment methods are disabled
<img width="384" alt="Screen Shot 2021-06-21 at 1 08 00 PM" src="https://user-images.githubusercontent.com/4083656/122837335-c4060680-d2a8-11eb-85da-afb7cc7975eb.png">
- Successfully pledge to a project
- Successfully update a pledge
- Successfully change your payment method
- Successfully change the reward tier
- Successfully use 3DS visa cards to make a payment

|  |  |

# 📋 QA
- Add new payment method, from settings -> account -> payment methods
- Add new payment method, from pledge screen -> new payment method
- make sure you can add all the supported card types `AMERICAN_EXPRESS,DINERS_CLUB,DISCOVER,JCB,MASTERCARD,UNIONPAY,VISA`, here you can see tes cards for all of those types. [Stripe Test Cards](https://stripe.com/docs/testing#cards)
- make sure you can do payments with any of the previous type of payments 
- make sure you can do payments with the 3DS visa types, here some flow samples and test cards https://kickstarter.atlassian.net/wiki/spaces/PAYM/pages/1083671123/Known+UI+issues
- make sure if you pledge to a non USA project you cannot use any of these payment methods: `DINERS_CLUB,DISCOVER,JCB,UNIONPAY,`
- Successfully pledge to a project
- Successfully update a pledge
- Successfully change your payment method
- Successfully change the reward tier

# Story 📖
[NT-2059](https://kickstarter.atlassian.net/browse/NT-2059)
